File: exercism/python/protein-translation/protein_translation.py
import sys

#  Returns Protein Translation
# Better solution
# https://exercism.io/tracks/python/exercises/protein-translation/solutions/61cb3d6d0b7e4e79b5cb5c8496de8783
def proteins(strand):
    ret = []

    map = dict()

    map['AUG'] = 'Methionine'
    for i in ['UUU', 'UUC']:
        map[i] = 'Phenylalanine'
    for i in ['UUA', 'UUG']:
        map[i] = 'Leucine'
    for i in ['UCU', 'UCC', 'UCA', 'UCG']:
        map[i] = 'Serine'
    for i in ['UAU', 'UAC']:
        map[i] = 'Tyrosine'
    for i in ['UGU', 'UGC']:
        map[i] = 'Cysteine'
    map['UGG'] = 'Tryptophan'
    for i in ['UAA', 'UAG', 'UGA']:
        map[i] = 'STOP'


    for i in range (0,len(strand), 3):
        part = strand[i:i+3]
        value = map.get(part)
        if value:
            if value == 'STOP':
                return ret
            ret.append(value)
    return ret


if __name__ == "__main__":
    print(proteins(sys.argv[1]))

File: exercism/go/protein-translation/protein_translation.go
package protein

import "errors"

// ErrStop : stop protein
var ErrStop = errors.New("Encountered STOP protein")

// ErrInvalidBase : invalid string
var ErrInvalidBase = errors.New("Encountered invalid base")

// var (
// 	// ErrStop : stop protein
// 	ErrStop = constError("ErrStop")

// 	// ErrInvalidBase : invalid string
// 	ErrInvalidBase = constError("ErrInvalidBase")
// )

// // for creation of const Error string
// type constError string

// func (err constError) Error() string {
// 	return string(err)
// }

// FromCodon translate condon to protein
func FromCodon(input string) (output string, e error) {
	switch input {
	case "AUG":
		return "Methionine", nil
	case "UUU", "UUC":
		return "Phenylalanine", nil
	case "UUA", "UUG":
		return "Leucine", nil
	case "UCU", "UCC", "UCA", "UCG":
		return "Serine", nil
	case "UAU", "UAC":
		return "Tyrosine", nil
	case "UGU", "UGC":
		return "Cysteine", nil
	case "UGG":
		return "Tryptophan", nil
	case "UAA", "UAG", "UGA":
		return "", ErrStop
	default:
		return "", ErrInvalidBase
	}
}

// FromRNA translate condon combo to protein array
func FromRNA(input string) ([]string, error) {
	var output []string
	for i := 0; i < len(input); i += 3 {
		c, err := FromCodon(input[i : i+3])
		if err != nil {
			if err == ErrStop {
				err = nil
			}
			return output, err
		}
		output = append(output, c)
	}

	return output, nil
}
